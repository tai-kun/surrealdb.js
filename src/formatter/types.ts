import type { Uint8ArrayLike } from "@tai-kun/surrealdb/types";
import type { StatefulPromise } from "@tai-kun/surrealdb/utils";
import type EncodedCBOR from "./encoded-cbor";
import type EncodedJSON from "./encoded-json";

export type Data = typeof globalThis extends
  { Buffer: new(...args: any) => infer Buff }
  ? string | ArrayBuffer | Uint8Array | Buff | Buff[]
  : string | ArrayBuffer | Uint8Array;

// dprint-ignore
export type EncodedData = string | Uint8ArrayLike

export type DecodingContext =
  | { name: "fetch"; length: number }
  | never; // 追加

export type DecodingStrategy = "sync" | "stream";

export interface Formatter {
  //////////////
  //  config  //
  //////////////

  readonly contentType: string;
  readonly wsProtocols: string[];

  //////////////
  // encoding //
  //////////////

  readonly toEncoded?: <T>(data: T) => EncodedCBOR<T> | EncodedJSON<T>;
  readonly encodeSync: (data: unknown) => EncodedData;
  // readonly encodeStream?: () => PromiseLike<unknown>;

  //////////////
  // decoding //
  //////////////

  readonly decodeSync: (data: Data) => unknown;
  readonly decodeStream?: (
    data: ReadableStream<Uint8Array>,
    signal: AbortSignal,
  ) => StatefulPromise<unknown>;
  /**
   * @experimental
   */
  readonly decodingStrategy?: (ctx: DecodingContext) => DecodingStrategy;
}
