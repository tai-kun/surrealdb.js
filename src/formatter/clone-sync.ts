import type { Formatter } from "./types";

export default function cloneSync<T>(formatter: Formatter, data: T): T {
  const encoded = formatter.encodeSync(data);
  const deocded = formatter.decodeSync(encoded);

  return deocded as T;
}
