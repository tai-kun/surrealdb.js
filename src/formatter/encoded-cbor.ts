import type { ToCBOR } from "@tai-kun/surrealdb/cbor";
import Encoded from "./encoded";

export default class EncodedCBOR<T = unknown, B = unknown> extends Encoded<T>
  implements ToCBOR
{
  readonly toCBOR: ToCBOR["toCBOR"];

  constructor(
    bindings: B,
    toCBOR: (
      this: B,
      ...args: Parameters<ToCBOR["toCBOR"]>
    ) => ReturnType<ToCBOR["toCBOR"]>,
  ) {
    super();
    this.toCBOR = toCBOR.bind(bindings);
  }
}
