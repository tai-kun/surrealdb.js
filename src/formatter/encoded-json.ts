import Encoded from "./encoded";

export default class EncodedJSON<T = unknown, B = unknown> extends Encoded<T> {
  readonly toJSON: () => unknown; // toRawJSON に対応していない場合に備える
  readonly toRawJSON?: () => string;

  constructor(
    bindings: B,
    toJSON: (this: B) => unknown, // toRawJSON に対応していない場合に備える
    toRawJSON?: ((this: B) => string) | undefined,
  ) {
    super();
    this.toJSON = toJSON.bind(bindings);

    if (toRawJSON) {
      this.toRawJSON = toRawJSON.bind(bindings);
    }
  }
}
