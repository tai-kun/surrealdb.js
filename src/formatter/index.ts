export type * from "./clone-sync";
export { default as cloneSync } from "./clone-sync";

export type * from "./encoded-cbor";
export { default as EncodedCBOR } from "./encoded-cbor";

export type * from "./encoded-json";
export { default as EncodedJSON } from "./encoded-json";

export type * from "./encoded";
export { default as Encoded } from "./encoded";

export type * from "./types";
