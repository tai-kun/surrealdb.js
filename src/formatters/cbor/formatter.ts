import {
  CONTINUE,
  decode,
  type DecodeOptions as CborDecodeOptions,
  decodeStream,
  encode,
  type EncodeOptions as CborEncodeOptions,
  Tagged,
} from "@tai-kun/surrealdb/cbor";
import type {
  BoundExcludedSource,
  BoundIncludedSource,
  DatetimeSource,
  DecimalSource,
  DurationSource,
  FutureSource,
  GeometryLineSource,
  GeometryMultiLineSource,
  GeometryMultiPointSource,
  GeometryMultiPolygonSource,
  GeometryPointSource,
  GeometryPolygonSource,
  // RangeSource,
  TableSource,
  ThingSource,
  UuidSource,
} from "@tai-kun/surrealdb/data-types/decode-only";
import {
  CBOR_TAG_BOUND_EXCLUDED,
  CBOR_TAG_BOUND_INCLUDED,
  CBOR_TAG_DATETIME,
  CBOR_TAG_DECIMAL,
  CBOR_TAG_DURATION,
  CBOR_TAG_FUTURE,
  CBOR_TAG_GEOMETRY_COLLECTION,
  CBOR_TAG_GEOMETRY_LINE,
  CBOR_TAG_GEOMETRY_MULTILINE,
  CBOR_TAG_GEOMETRY_MULTIPOINT,
  CBOR_TAG_GEOMETRY_MULTIPOLYGON,
  CBOR_TAG_GEOMETRY_POINT,
  CBOR_TAG_GEOMETRY_POLYGON,
  CBOR_TAG_NONE,
  CBOR_TAG_RANGE,
  CBOR_TAG_RECORDID,
  CBOR_TAG_TABLE,
  CBOR_TAG_UUID,
} from "@tai-kun/surrealdb/data-types/encodable";
import { SurrealTypeError } from "@tai-kun/surrealdb/errors";
import {
  type Data,
  EncodedCBOR,
  type Formatter,
} from "@tai-kun/surrealdb/formatter";
import type { Uint8ArrayLike } from "@tai-kun/surrealdb/types";
import {
  isArrayBuffer,
  type StatefulPromise,
  utf8,
} from "@tai-kun/surrealdb/utils";

const hasBuffer = typeof Buffer !== "undefined";

export interface CborDataTypes {
  readonly Uuid: new(source: UuidSource) => any;
  readonly Range: new(source: readonly [any, any]) => any;
  readonly Table: new(source: TableSource) => any;
  readonly Thing: new(source: ThingSource) => any;
  readonly Future: new(source: FutureSource) => any;
  readonly Decimal: new(source: DecimalSource) => any;
  readonly Datetime: new(source: DatetimeSource) => any;
  readonly Duration: new(source: DurationSource) => any;
  readonly GeometryLine: new(source: GeometryLineSource) => any;
  readonly BoundIncluded: new(source: BoundIncludedSource) => any;
  readonly BoundExcluded: new(source: BoundExcludedSource) => any;
  readonly GeometryPoint: new(source: GeometryPointSource) => any;
  readonly GeometryPolygon: new(source: GeometryPolygonSource) => any;
  readonly GeometryMultiLine: new(source: GeometryMultiLineSource) => any;
  readonly GeometryMultiPoint: new(source: GeometryMultiPointSource) => any;
  readonly GeometryCollection: new(source: readonly any[]) => any;
  readonly GeometryMultiPolygon: new(source: GeometryMultiPolygonSource) => any;
}

export interface EncodeOptions extends CborEncodeOptions {}

export interface DecodeOptions extends CborDecodeOptions {
  readonly streamThreshold?: number | undefined;
}

export interface CborFormatterOptions extends CborDataTypes {
  readonly encode?: EncodeOptions | undefined;
  readonly decode?: DecodeOptions | undefined;
}

export default class CborFormatter implements Formatter {
  protected dataTypes: CborDataTypes;
  protected encodeOptions: EncodeOptions;
  protected decodeOptions: DecodeOptions;

  contentType = "application/cbor";
  wsProtocols = ["cbor"];

  constructor(options: CborFormatterOptions) {
    const {
      encode = {},
      decode = {},
      Uuid,
      Range,
      Table,
      Thing,
      Future,
      Decimal,
      Datetime,
      Duration,
      GeometryLine,
      BoundIncluded,
      BoundExcluded,
      GeometryPoint,
      GeometryPolygon,
      GeometryMultiLine,
      GeometryCollection,
      GeometryMultiPoint,
      GeometryMultiPolygon,
    } = options;
    this.dataTypes = {
      Uuid,
      Range,
      Table,
      Thing,
      Future,
      Decimal,
      Datetime,
      Duration,
      GeometryLine,
      BoundIncluded,
      BoundExcluded,
      GeometryPoint,
      GeometryPolygon,
      GeometryMultiLine,
      GeometryCollection,
      GeometryMultiPoint,
      GeometryMultiPolygon,
    };
    this.encodeOptions = encode;
    this.decodeOptions = Object.assign({
      reviver: {
        tagged: (t: Tagged<any>) => {
          switch (t.tag) {
            case CBOR_TAG_NONE:
              return undefined;

            case CBOR_TAG_TABLE:
              return new Table(t.value);

            case CBOR_TAG_RECORDID:
              return new Thing(t.value);

            case CBOR_TAG_DECIMAL:
              return new Decimal(t.value);

            case CBOR_TAG_DATETIME:
              return new Datetime(t.value);

            case CBOR_TAG_DURATION:
              return new Duration(t.value);

            case CBOR_TAG_FUTURE:
              return new Future(t.value);

            case CBOR_TAG_UUID:
              return new Uuid(t.value);

            case CBOR_TAG_RANGE:
              return new Range(t.value);

            case CBOR_TAG_BOUND_INCLUDED:
              return new BoundIncluded(t.value);

            case CBOR_TAG_BOUND_EXCLUDED:
              return new BoundExcluded(t.value);

            case CBOR_TAG_GEOMETRY_POINT:
              return new GeometryPoint(t.value);

            case CBOR_TAG_GEOMETRY_LINE:
              return new GeometryLine(t.value);

            case CBOR_TAG_GEOMETRY_POLYGON:
              return new GeometryPolygon(t.value);

            case CBOR_TAG_GEOMETRY_MULTIPOINT:
              return new GeometryMultiPoint(t.value);

            case CBOR_TAG_GEOMETRY_MULTILINE:
              return new GeometryMultiLine(t.value);

            case CBOR_TAG_GEOMETRY_MULTIPOLYGON:
              return new GeometryMultiPolygon(t.value);

            case CBOR_TAG_GEOMETRY_COLLECTION:
              return new GeometryCollection(t.value);

            default:
              return CONTINUE;
          }
        },
      },
    }, decode);
  }

  toEncoded<TData = unknown>(data: TData): EncodedCBOR<TData> {
    if (data instanceof EncodedCBOR) {
      return data;
    }

    return new EncodedCBOR(
      this.encodeSync(data),
      function toCBOR(w) {
        w.writeBytes(this);
      },
    );
  }

  encodeSync(data: unknown): Uint8Array {
    return encode(data, this.encodeOptions);
  }

  decodeSync(data: Data): unknown {
    return decode(toEncodedData(data), this.decodeOptions);
  }

  decodeStream(
    data: ReadableStream<Uint8Array>,
    signal: AbortSignal,
  ): StatefulPromise<unknown> {
    return decodeStream(data, Object.assign({ signal }, this.decodeOptions));
  }

  // /** @experimental */
  // decodingStrategy(ctx: DecodingContext): DecodingStrategy {
  //   // TODO(tai-kun): 8 KiB は適当。要調整。
  //   if (ctx.length > (this.decodeOptions.streamThreshold || 8192)) {
  //     return "stream";
  //   }

  //   return "sync";
  // }
}

function toEncodedData(data: Data): Uint8ArrayLike {
  switch (true) {
    case data instanceof Uint8Array:
      return hasBuffer && data instanceof Buffer
        ? new Uint8Array(data)
        : data as Uint8Array;

    case isArrayBuffer(data):
      return new Uint8Array(data);

    case hasBuffer
      && Array.isArray(data)
      && data.every(b => Buffer.isBuffer(b)):
      return Buffer.concat(data);

    case typeof data === "string":
      return utf8.encode(data);

    default:
      throw new SurrealTypeError(
        ["String", "Buffer", "ArrayBuffer", "Uint8Array", "Buffer[]"],
        data,
      );
  }
}
