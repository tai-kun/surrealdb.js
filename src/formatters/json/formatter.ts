import {
  JsonUnsafeMapKeyError,
  SurrealTypeError,
} from "@tai-kun/surrealdb/errors";
import {
  cloneSync,
  type Data,
  EncodedJSON,
  type Formatter,
} from "@tai-kun/surrealdb/formatter";
import { isArrayBuffer, utf8 } from "@tai-kun/surrealdb/utils";

// https://github.com/fastify/secure-json-parse
const __PROTO__REGEX =
  /"(?:_|\\u005[Ff])(?:_|\\u005[Ff])(?:p|\\u0070)(?:r|\\u0072)(?:o|\\u006[Ff])(?:t|\\u0074)(?:o|\\u006[Ff])(?:_|\\u005[Ff])(?:_|\\u005[Ff])"\s*:/;
const CONSTRUCTOR_REGEX =
  /"(?:c|\\u0063)(?:o|\\u006[Ff])(?:n|\\u006[Ee])(?:s|\\u0073)(?:t|\\u0074)(?:r|\\u0072)(?:u|\\u0075)(?:c|\\u0063)(?:t|\\u0074)(?:o|\\u006[Ff])(?:r|\\u0072)"\s*:/;

// TODO(tai-kun): isSafeMapKey, isSafeObjectKey をオプションに追加
export default class JsonFormatter implements Formatter {
  contentType = "application/json";
  wsProtocols = ["json"];

  encodeSync(data: unknown): string {
    return JSON.stringify(data);
  }

  decodeSync(data: Data): unknown {
    const json = toString(data);

    if (__PROTO__REGEX.test(json)) {
      throw new JsonUnsafeMapKeyError("__proto__");
    }

    if (CONSTRUCTOR_REGEX.test(json)) {
      throw new JsonUnsafeMapKeyError("constructor");
    }

    return JSON.parse(json);
  }

  cloneSync<T>(data: T): T {
    switch (typeof data) {
      case "string":
      case "number":
      case "boolean":
        return data;

      default:
        switch (data) {
          case null:
          case undefined:
            return data;

          default:
            return cloneSync(this, data);
        }
    }
  }

  toEncoded<TData>(data: TData): EncodedJSON<TData> {
    if (data instanceof EncodedJSON) {
      return data;
    }

    return new EncodedJSON(
      {
        data: this.cloneSync(data),
      },
      function toJSON() {
        return this.data;
      },
      // TODO(tai-kun): JSON フォーマッターの事前エンコーディングを実装する。
      // function toRawJSON() {
      //   return (this.json ??= JSON.stringify(this.data));
      // },
    );
  }
}

function toString(data: Data): string {
  switch (true) {
    case typeof data === "string":
      return data;

    case data instanceof Uint8Array || isArrayBuffer(data):
      return utf8.decode(data);

    case typeof Buffer !== "undefined"
      && Array.isArray(data)
      && data.every(b => Buffer.isBuffer(b)):
      return utf8.decode(Buffer.concat(data));

    default:
      throw new SurrealTypeError(
        ["String", "Buffer", "ArrayBuffer", "Uint8Array", "Buffer[]"],
        data,
      );
  }
}
