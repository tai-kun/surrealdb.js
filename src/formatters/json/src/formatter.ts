import { SurrealTypeError } from "@tai-kun/surrealdb/errors";
import {
  cloneSync,
  type Data,
  EncodedJSON,
  type Formatter,
} from "@tai-kun/surrealdb/formatter";
import { isArrayBuffer, utf8 } from "@tai-kun/surrealdb/utils";

export default class JsonFormatter implements Formatter {
  mimeType = "application/json";

  encodeSync(data: unknown): string {
    return JSON.stringify(data);
  }

  decodeSync(data: Data): unknown {
    return JSON.parse(toString(data));
  }

  cloneSync<T>(data: T): T {
    switch (typeof data) {
      case "string":
      case "number":
      case "boolean":
        return data;

      default:
        switch (data) {
          case null:
          case undefined:
            return data;

          default:
            return cloneSync(this, data);
        }
    }
  }

  toEncoded<T>(data: T): EncodedJSON<T> {
    if (data instanceof EncodedJSON) {
      return data;
    }

    return new EncodedJSON(
      {
        data: this.cloneSync(data),
      },
      function toJSON() {
        return this.data;
      },
      // TODO(tai-kun): JSON フォーマッターの事前エンコーディングを実装する。
      // function toRawJSON() {
      //   return (this.json ??= JSON.stringify(this.data));
      // },
    );
  }
}

function toString(data: Data): string {
  switch (true) {
    case typeof data === "string":
      return data;

    case data instanceof Uint8Array || isArrayBuffer(data):
      return utf8.decode(data);

    case typeof Buffer !== "undefined"
      && Array.isArray(data) && data.every(b => Buffer.isBuffer(b)):
      return utf8.decode(Buffer.concat(data));

    default:
      throw new SurrealTypeError(
        ["String", "Buffer", "ArrayBuffer", "Uint8Array", "Buffer[]"],
        data,
      );
  }
}
