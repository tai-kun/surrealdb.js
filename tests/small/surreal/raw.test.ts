import { surql, Thing } from "@tai-kun/surrealdb";
import type { Writer } from "@tai-kun/surrealdb/cbor";
import { EncodedCBOR } from "@tai-kun/surrealdb/formatter";
import { afterEach, expect, test, vi } from "vitest";

const writeBytes = vi.fn();
const writer = { writeBytes } as unknown as Writer;

afterEach(() => {
  writeBytes.mockClear();
});

test("文字列を変数ではなく SurrealQL に埋め込む", () => {
  const ReturnQuery = surql`${surql.raw("RETURN 0")};`;

  expect(ReturnQuery.text).toBeInstanceOf(EncodedCBOR);
  (ReturnQuery.text as EncodedCBOR).toCBOR(writer);
  expect(writeBytes.mock.calls).toStrictEqual([
    [
      // `RETURN 0;`
      new Uint8Array([
        0b011_01001, // mt: 3, ai: 9
        0x52, // `R`
        0x45, // `E`
        0x54, // `T`
        0x55, // `U`
        0x52, // `R`
        0x4e, // `N`
        0x20, // ` `
        0x30, // `0`
        0x3b, // `;`
      ]),
    ],
  ]);
});

test(".toSurql() が返す文字列を SurrealQL に埋め込む", () => {
  const ReturnQuery = surql`${surql.raw({ toSurql: () => "RETURN 0" })};`;

  expect(ReturnQuery.text).toBeInstanceOf(EncodedCBOR);
  (ReturnQuery.text as EncodedCBOR).toCBOR(writer);
  expect(writeBytes.mock.calls).toStrictEqual([
    [
      // `RETURN 0;`
      new Uint8Array([
        0b011_01001, // mt: 3, ai: 9
        0x52, // `R`
        0x45, // `E`
        0x54, // `T`
        0x55, // `U`
        0x52, // `R`
        0x4e, // `N`
        0x20, // ` `
        0x30, // `0`
        0x3b, // `;`
      ]),
    ],
  ]);
});

test(".toString() が返す文字列を SurrealQL に埋め込む", () => {
  const ReturnQuery = surql`${surql.raw({ toString: () => "RETURN 0" })};`;

  expect(ReturnQuery.text).toBeInstanceOf(EncodedCBOR);
  (ReturnQuery.text as EncodedCBOR).toCBOR(writer);
  expect(writeBytes.mock.calls).toStrictEqual([
    [
      // `RETURN 0;`
      new Uint8Array([
        0b011_01001, // mt: 3, ai: 9
        0x52, // `R`
        0x45, // `E`
        0x54, // `T`
        0x55, // `U`
        0x52, // `R`
        0x4e, // `N`
        0x20, // ` `
        0x30, // `0`
        0x3b, // `;`
      ]),
    ],
  ]);
});

test(".toString() 明示的に定義されていないならエラー", () => {
  expect(() => surql.raw({})).toThrowErrorMatchingSnapshot();
});

test("ID ジェネレーターをエスケープしない", () => {
  const withGenId = new Thing("user", "uuid()");

  expect(String(withGenId)).toBe("user:⟨uuid()⟩");
  expect(String(surql.raw(withGenId))).toBe("user:uuid()");
});

test("ID ジェネレーターをエスケープせずにエンコードする", () => {
  const withGenId = new Thing("user", "uuid()");
  const ReturnQuery = surql`${surql.raw(withGenId)};`;

  expect(ReturnQuery.text).toBeInstanceOf(EncodedCBOR);
  (ReturnQuery.text as EncodedCBOR).toCBOR(writer);
  expect(writeBytes.mock.calls).toStrictEqual([
    [
      // `user:uuid();`
      new Uint8Array([
        0b011_01100, // mt: 3, ai: 12
        0x75, // `u`
        0x73, // `s`
        0x65, // `e`
        0x72, // `r`
        0x3a, // `:`
        0x75, // `u`
        0x75, // `u`
        0x69, // `i`
        0x64, // `d`
        0x28, // `(`
        0x29, // `)`
        0x3b, // `;`
      ]),
    ],
  ]);
});

test("数値を入力してエラー", () => {
  expect(() => surql.raw(0)).toThrowErrorMatchingSnapshot();
});
