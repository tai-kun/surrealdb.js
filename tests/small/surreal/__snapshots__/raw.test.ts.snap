// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`.toString() 明示的に定義されていないならエラー 1`] = `
{
  "actual": "object",
  "expected": "object",
  "message": "Expected object but got object.",
  "name": "SurrealTypeError",
}
`;

exports[`数値を入力してエラー 1`] = `
{
  "actual": "number",
  "expected": "string | object",
  "message": "Expected string | object but got number.",
  "name": "SurrealTypeError",
}
`;
